text/3637-guard-patterns.md

```diff
@@ -250,6 +250,42 @@ If guards can only appear immediately within or-patterns, then either
 
 This can also be seen as a special case of the previous argument, as pattern macros fundamentally assume that patterns can be built out of composable, local pieces.
 
+## Deref and Const Patterns Must Be Pure, But Not Guards
+
+It may seem odd that we explicitly require `deref!` and const patterns to use pure `Deref` and `PartialEq` implementations, respectively, but allow arbitrary side effects in guards. The ultimate reason for this is that, unlike `deref!` and const patterns, guard patterns are always refutable.
```

Comment:
Since `deref!` patterns aren't part of this RFC and aren't part of the language, let's cite the specific proposal for these.

And let's please separate out the discussion of what is in the language from what is still a proposal.  E.g., "Today, consts in patterns require... Similarly, there's a proposal (*link*) for `deref!` patterns that would require...".

Reply:
I've rewritten the section to clarify this, thanks.